src/browser-polyfill.js

```diff
@@ -396,36 +396,96 @@ if (typeof browser === "undefined") {
+          promise.then(msg => {
+            // send the message value.
+            sendResponse(msg);
+          }, error => {
+            console.error(error);
+            // Send a JSON representation of the error if the rejected value
+            // is an instance of error, or the object itself otherwise.
+            const isErrorInstance = error instanceof Error;
+            sendResponse({
+              __webExtensionPolyfillReject__: {isErrorInstance},
+              error: !isErrorInstance ? error : {
+                message: error.message,
+                fileName: error.fileName,
+                lineNumber: error.lineNumber,
+              },
+            });
+          }).catch(err => {
+            // Print an error on the console if unable to send the response.
+            console.error(err);
+          });
+        };
+
         // If the listener returned a Promise, send the resolved value as a
         // result, otherwise wait the promise related to the wrappedSendResponse
         // callback to resolve and send it as a response.
         if (isResultThenable) {
-          result.then(sendResponse, error => {
-            console.error(error);
-            // TODO: the error object is not serializable and so for now we just send
-            // `undefined`. Nevertheless, as being discussed in #97, this is not yet
-            // providing the expected behavior (the promise received from the sender should
-            // be rejected when the promise returned by the listener is being rejected).
-            sendResponse(undefined);
-          });
+          sendPromisedResult(result);
         } else {
-          sendResponsePromise.then(sendResponse, error => {
-            console.error(error);
-            // TODO: same as above, we are currently sending `undefined` in this scenario
-            // because the error oject is not serializable, but it is not yet the behavior
-            // that this scenario should present.
-            sendResponse(undefined);
-          });
+          sendPromisedResult(sendResponsePromise);
         }
 
         // Let Chrome know that the listener is replying.
         return true;
       };
     });
 
+    const wrappedSendMessageCallback = ({reject, resolve, unexpectedError}, reply) => {
+      if (reply && reply.__webExtensionPolyfillReject__) {
+        const {error} = reply;
+        // Convert the JSON representation of the error back to an Error instance,
+        // otherwise just reject with the error value as it is.
+        if (reply.__webExtensionPolyfillReject__.isErrorInstance) {
+          reject(new Error(error.message, error.fileName, error.lineNumber));
+        } else {
+          // If the original error wasn't an error instance, let's try to keep
+          // the behavior as close as possible to the Firefox behavior, e.g.
+          // print the error object as received in the console and then reject
+          // using a generic error object.
```

Comment:
This is not Firefox's behavior, and even if it was, then it would be too implementation-specific.

Test case (Firefox Nightly 61, buildID 20180504220115):
```javascript
browser.runtime.onMessage.addListener(()=>Promise.reject({message:"what"}))
```

```javascript
// If doing this, no error printed:
browser.runtime.sendMessage('ee').catch(()=>{});
// If doing this, then error is printed (because it's an uncaught promise rejection):
browser.runtime.sendMessage('ee');
```

Reply:
:+1: I was also pretty unsure about these console errors, to be fair I'm still a bit worried that the stack traces that the extension is going to be able to access directly from their code could be not that helpful (at least in some of the scenarios), but I've removed all the redundant console messages for now.